File: Solutions/94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
 

//recursive 
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        inorder(res,root);
        return res;
    }
private:
    void inorder(vector<int>& res,TreeNode* root){
        if(root==nullptr)
            return;
        inorder(res,root->left);
        res.push_back(root->val);
        inorder(res,root->right);
    }
};



//iteration stack
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> stk;
        while(!stk.empty()||root){
            while(root){
                stk.push(root);
                root=root->left;
            }
            root=stk.top();
            stk.pop();
            res.push_back(root->val);
            root=root->right;
        }
        return res;
    }
};

File: Solutions/94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//recursive
func inorderTraversal(root *TreeNode) []int {
    var res []int
    inorder(root,&res)
    return res
}

func inorder(root *TreeNode,res *[]int){
    if(root!=nil){
        inorder(root.Left,res)
        *res=append(*res,root.Val)
        inorder(root.Right,res)
    }
}

//iteration ,stack
func inorderTraversal(root *TreeNode) []int {
    var stk []*TreeNode
    var res []int
    for len(stk)>0 || root!=nil{
        for root!=nil{
            stk=append(stk,root)
            root=root.Left
        }
        top:=len(stk)-1  //top index
        root=stk[top]  
        stk=stk[:top]    //pop
        res=append(res,root.Val)
        root=root.Right
    }
    return res
}